### ISSUE TITLE
No qualifying bean of type org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizedClientRepository

### ISSUE THREAD

**Issue by abhimanuhandoo** [oktadeveloper/okta-legacy-with-cloud-gateway-example#3]
I am getting this error while trying to build application as war and then deploy it on tomcat .
Below is the trace:

```
	Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.springframework.security.oauth2.client.web.server.ServerOAuth2AuthorizedClientRepository' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}
		at org.springframework.beans.factory.support.DefaultListableBeanFactory.raiseNoMatchingBeanFound(DefaultListableBeanFactory.java:1678)
		at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1237)
		at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1191)
		at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:857)
		at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:760)
```

**Comment by mraible** [comment]
Does it run if you use `./mvnw spring-boot:run`? This particular example is not designed to run in Tomcat as a WAR. See [Deploy a Spring Boot Application into Tomcat](https://developer.okta.com/blog/2019/04/16/spring-boot-tomcat) to see how to do that.

**Comment by abhimanuhandoo** [comment]
Thanks @mraible  for replying. I am able to run it using the mvn spring-boot:run command.
I checked the URL mentioned (https://developer.okta.com/blog/2019/04/16/spring-boot-tomcat).
However when I deployed the war in tomcat I got error like :
```
Caused by: java.lang.IllegalArgumentException: class org.springframework.boot.cloud.CloudFoundryVcapEnvironmentPostProcessor is not assignable to interface org.springframework.context.ApplicationListener
at org.springframework.util.Assert.assignableCheckFailed(Assert.java:673)
at org.springframework.util.Assert.isAssignable(Assert.java:604)
at org.springframework.util.Assert.isAssignable(Assert.java:635)
at org.springframework.boot.SpringApplication.createSpringFactoriesInstances(SpringApplication.java:443)
... 57 more
```

Post getting this error , I modified the pom.xml and added below dependency to resolve above issue
in properties tag added 
```
<spring-cloud.version>Greenwich.SR2</spring-cloud.version>
```
and in dependency added 
```
<dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-oauth2</artifactId>
</dependency>
```
and below as well
```
    <dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework.cloud</groupId>
				<artifactId>spring-cloud-dependencies</artifactId>
				<version>${spring-cloud.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>
````

After that re-created the war and now getting below error trace
```
Caused by: java.lang.IllegalArgumentException: Unable to instantiate factory class: org.springframework.boot.env.EnvironmentPostProcessor
                at org.springframework.core.io.support.SpringFactoriesLoader.instantiateFactory(SpringFactoriesLoader.java:167)
                at org.springframework.core.io.support.SpringFactoriesLoader.loadFactories(SpringFactoriesLoader.java:104)
                at org.springframework.boot.context.config.ConfigFileApplicationListener.loadPostProcessors(ConfigFileApplicationListener.java:183)
                at org.springframework.boot.context.config.ConfigFileApplicationListener.onApplicationEnvironmentPreparedEvent(ConfigFileApplicationListener.java:174)
                at org.springframework.boot.context.config.ConfigFileApplicationListener.onApplicationEvent(ConfigFileApplicationListener.java:166)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:127)
                at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:76)
                at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:53)
                at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:342)
                at org.springframework.boot.SpringApplication.run(SpringApplication.java:305)
                at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:140)
                at org.springframework.cloud.bootstrap.BootstrapApplicationListener.bootstrapServiceContext(BootstrapApplicationListener.java:203)
                at org.springframework.cloud.bootstrap.BootstrapApplicationListener.onApplicationEvent(BootstrapApplicationListener.java:114)
                at org.springframework.cloud.bootstrap.BootstrapApplicationListener.onApplicationEvent(BootstrapApplicationListener.java:71)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
                at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:127)
                at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:76)
                at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:53)
                at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:342)
                at org.springframework.boot.SpringApplication.run(SpringApplication.java:305)
                at org.springframework.boot.web.servlet.support.SpringBootServletInitializer.run(SpringBootServletInitializer.java:151)
                at org.springframework.boot.web.servlet.support.SpringBootServletInitializer.createRootApplicationContext(SpringBootServletInitializer.java:131)
                at org.springframework.boot.web.servlet.support.SpringBootServletInitializer.onStartup(SpringBootServletInitializer.java:91)
                at org.springframework.web.SpringServletContainerInitializer.onStartup(SpringServletContainerInitializer.java:172)
                at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5128)
                at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
                ... 45 more
        Caused by: java.lang.IllegalArgumentException: Class [org.springframework.cloud.gateway.config.GatewayEnvironmentPostProcessor] is not assignable to [org.springframework.boot.env.EnvironmentPostProcessor]
                at org.springframework.core.io.support.SpringFactoriesLoader.instantiateFactory(SpringFactoriesLoader.java:162)
                ... 75 more
```

Request you to please help me with this issue. Thanks!!

**Comment by mraible** [comment]
@abhimanuhandoo I was able to reproduce your issue in https://github.com/oktadeveloper/okta-legacy-with-cloud-gateway-example/pull/4. I've asked @bdemers to take a look.

**Comment by bdemers** [comment]
I thought this was going to be an easy one :) a missing servlet related dependency. (Spring automatically loads a servlet or reactive based implementation depending on what is on the classpath)

That wasn't the case, I had forgotten about the fact that Spring Cloud Gateway does NOT work a servlet container (as it requires Netty). See the note in the [Spring Cloud Gateway docs](https://cloud.spring.io/spring-cloud-gateway/reference/html/)

TL;DR - Spring Cloud Gateway needs to run in a standalone jar.

**Comment by bdemers** [comment]
@abhimanuhandoo,
Note that the "legacy" application (or any OAuth Resource Server) referenced in this post can be deployed as a WAR if needed)

**Comment by abhimanuhandoo** [comment]
Thanks @mraible  and @bdemers . It works as a standalone jar. However thought it might work as in a servlet container.

**Comment by abhimanuhandoo** [comment]
> @abhimanuhandoo,
> Note that the "legacy" application (or any OAuth Resource Server) referenced in this post can be deployed as a WAR if needed)

Do we have any straight forward approach to integrate okta oauth 2.0 with legacy application( servlet based). As noted I have to use war only and do not want to change my existing URL for the application for integration purpose.

**Comment by bdemers** [comment]
There are many options depending on what you want to do.  You could use the Spring Cloud Gateway option without changing your URLs, the gateway just sits in front of your application handling the inbound requests.  (This would be same for other gateway options, with Nginx, or any other third party gateway.

These are usually the first type solution I recommend because it's usually a low touch way to update existing applications to use OAuth, and if/when you update your existing application to more modern frameworks, the interaction between the gateway and app doesn't need to change.

That said, there are other ways to add OAuth into your application.  It depends on which frameworks you are using, but you may be able to find an existing OAuth library you can integrate into your app.
You could also use something like Spring Security in a non Spring application (assuming you do not have library/version conflicts)

Can you tell us a little about your stack, maybe we can suggest a option?
